EgbaFrank/Portfolio-project
# Setup RESTful API with Swagger Documentation

### Description:
This PR introduces the initial setup of the RESTful API for the Grocery Hub platform, alongside integration with Swagger for comprehensive API documentation.

### Key Features:

1. Flask RESTful API:
    - The API structure follows RESTful principles, with separate routes for each resource.
    - Models include Category, Product, ShopList, and more, which interact with the database to handle CRUD operations.
    - Error handling and response codes are standardized across the application.

2. Swagger Documentation:
    - Integrated Swagger UI for interactive API documentation.
    - Swagger documentation is written in YAML format and auto-generated for all API routes using swag_from.
    - Allows for easy testing and interaction with the API via the browser.

### Changes Made:

1. API Routes:
  Implemented key API routes for managing resources like categories, products, and shopping lists:
    - GET /categories: Retrieve all categories.
    - POST /categories: Create a new category.
    - PUT /categories/<category_id>: Update an existing category.
    - GET /products: Retrieve all products (with pagination, sorting, filtering).
    - POST /products: Create a new product.

2. Swagger Configuration:
    - Configured Swagger UI to generate live documentation.
    - Set up external YAML files to store route and schema definitions for modular and maintainable documentation.
    - Swagger UI is accessible at /apidocs/.

3. Modular API Documentation:
    - Defined each method and route in separate YAML files under api_docs/{model}.
    - Routes and endpoints are well-documented, with clear examples of request/response formats, required parameters, and possible error messages.

4. Pagination and Filtering:
    - The API now supports pagination, sorting, and filtering for listing routes (e.g., GET /products).
    - Query parameters such as offset, limit, sort, and order are supported for customizable data retrieval.

### How to Test:

1. Start the API server:
- Ensure all dependencies are installed, including Flask and Swagger-related modules.
- Run the server using `python3 -m api.v1.app` from the root of the repository

2. Access the API:
Use an API client like Postman or Curl to interact with the API routes.
For example:

- GET /categories
- POST /categories
- PUT /categories/<category_id>

3. Access the API documentation:
- Visit /apidocs/ in the browser to view and interact with the Swagger UI.
- Test endpoints directly from the Swagger UI interface.

### Next Steps:

- Add authentication and authorization middleware for secured access to API routes.
- Expand API documentation to cover error responses and advanced features like product search with filters.
- Implement more advanced features like rate-limiting and caching for optimization.

### Related Issues:
Closes #9  (Setting up the RESTful API)

**Notes for Reviewers**:

- Ensure that the Swagger UI is functioning correctly and displays all routes with proper request/response formats.
- Check the modular documentation for reusability and clarity.
- Review the API logic for route handling and error handling consistency.
